<!DOCTYPE html>
<html>

<head>
<link rel=stylesheet href=style.css />
<link rel=icon href=CZI-new-logo.png />
</head>

<body>

<h1>Current status</h1>

<p>
An initial release of the Shasta assembler is imminent and expected
in early May 2019.

<p>
As of April 2019, the Shasta assembler is functional and is
capable of performing a human assembly at typical coverage
in a few hours on a large machine with an appropriate configuration,
using as input long reads from Oxford Nanopore flowcells.

<p>
For example,
at coverage around 60x, an AWS  
<code>x1.32xlarge</code> instance, with 128 virtual processors and 1952 GB of memory,
completes the assembly in about 5 hours.
The compute cost of such an assembly is around $20 at AWS spot market or reserved prices.

<p>
The Shasta assembler, although not designed or optimized for smaller genomes,
can also assemble bacterial genomes in minutes on a typical laptop
(e. g. 1-2 minutes for an <i>E. coli</i> assembly). 

<p>
An evaluation of the quality of assembled sequence is in progress.
Early indications are that,
when comparing with alternative, less performant long read assemblers,
Shasta assembly quality is similar or better.
This evaluation is expected to be complete soon, and its results will
be made available.

<p>
Default Shasta assembly parameters are optimized for coverage around 60x, 
and require reconfiguration if using much lower or much higher coverage.

<p>
The Shasta assembler creates output in 
<a href=https://en.wikipedia.org/wiki/FASTA>FASTA</a> and 
<a href=https://github.com/GFA-spec/GFA-spec/blob/master/GFA1.md>GFA 1.0</a> formats.
It also includes an http server and a Python API 
that can be used to access and visualize
the data structures created during assembly.
This can be useful for testing or debugging,
or for studying assembler behavior at locations of interest.

<p>
The Shasta assembler is undergoing active development
and it is expected to improve in accuracy, performance,
and functionality.



</body>
</html>

